### PR TITLE
number input type validations

### LINKED ISSUE #19172: [meta] Implement value sanitizations for different input modes
#18262 laid the groundwork for this, but we're still missing value sanitizations for:

- [x] [Date type inputs](https://html.spec.whatwg.org/multipage/input.html#date-state-(type=date):value-sanitization-algorithm) (#19385)
- [x] [Month type inputs](https://html.spec.whatwg.org/multipage/input.html#month-state-(type=month):value-sanitization-algorithm) (#19385)
- [x] [Week type inputs](https://html.spec.whatwg.org/multipage/input.html#week-state-(type=week):value-sanitization-algorithm) (#19559)
- [x] [Time type inputs](https://html.spec.whatwg.org/multipage/input.html#time-state-(type=time):value-sanitization-algorithm) (#19379)
- [x] [Date-time type inputs](https://html.spec.whatwg.org/multipage/input.html#local-date-and-time-state-(type=datetime-local):value-sanitization-algorithm) (#19602)
- [x] [Number type inputs](https://html.spec.whatwg.org/multipage/input.html#number-state-(type=number):value-sanitization-algorithm) and [range type inputs](https://html.spec.whatwg.org/multipage/input.html#range-state-(type=range):value-sanitization-algorithm) (#19730, #19761)
- [x] [Color type inputs](https://html.spec.whatwg.org/multipage/input.html#color-state-(type=color):value-sanitization-algorithm) (#19330)

All of the missing algorithms above are expected to be implemented in this function: https://github.com/servo/servo/blob/a3f2fae561de26fefb0ac0d569ce2c7a2942eee5/components/script/dom/htmlinputelement.rs#L873-L888

The implementation for these value sanitizations are given by the spec, which I have linked accordingly. When each of these are implemented, some subtests under `tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/type-change-state.html` should be expected to PASS instead. Some other tests under the same directory may also pass as a result of a correct implementation.

If you want to help out with this, please comment on this issue or reach out on IRC (`#servo` in `irc.mozilla.org`), and we'll find one which is interesting to you; or, you can also file an issue and mention this one, and I'll get it assigned to you.

### PR DESCRIPTION
**Pull request by cursedquail**
I used rust's builtin float parser to implement this. Rust's parser is more permissive than what browsers support (in this case), so I added some code to handle those edge cases.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes are part of #19172

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19730)

### CODE REVIEW

**Comment by tigercosmos** [comment]
if you use "fix" keyword, the issue will be closed once the PR merged.
But this PR only do `number` part. There is still `range` one need to do.
In this case, we should just mention the issue than adding keyword `fix` or `close`

File: components/script/dom/bindings/str.rs [hunk 1]
```diff
@@ -303,6 +304,37 @@ impl DOMString {
         parse_week_string(&*self.0).is_ok()
     }
 
+    /// https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-floating-point-number
+    /// The rust default String::parse<f64>() is extremely close, and this just fills in the gaps.
+    pub fn is_valid_number_string(&self) -> bool {
+        let input = &self.0;
+        let parsed_result = input.parse::<f64>();
```
Comment (tigercosmos) [inline components/script/dom/bindings/str.rs]: It would be better
```
let parsed_result = match  input.parse::<f64>() {
// ... 
}
```

File: components/script/dom/bindings/str.rs [hunk 2]
```diff
@@ -303,6 +304,37 @@ impl DOMString {
...
+                }
+                // Another edge case. Rust interprets the literal "NaN" to be NaN
+                if val.is_nan() {
+                    return false;
+                }
+                // Finally, if the value starts with "+", it is invalid
+                if input.starts_with("+") {
+                    return false;
+                }
+                return true;
```
Comment (tigercosmos) [inline components/script/dom/bindings/str.rs]: I guess no need `return` here?

**Comment by tigercosmos** [comment]
If you not change anythings in `/tests`, we don't need to run `./mach update-manifest`

**Comment by tigercosmos** [comment]
```
Running the WPT lint on ./tests/wpt/mozilla/tests/...
Checking files for tidiness...
./components/script/dom/bindings/str.rs:307: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#valid-floating-point-number
./components/script/dom/bindings/str.rs:21: use statement is not in alphabetical order
	expected: std::str::{Bytes, FromStr}
	found: std::f64
```

File: components/script/dom/bindings/str.rs (hunk 1 above)
Comment (NLincoln) [reply components/script/dom/bindings/str.rs]: Agreed! Fixed.

File: components/script/dom/bindings/str.rs (hunk 2 above)
Comment (NLincoln) [reply components/script/dom/bindings/str.rs]: :+1: . I guess I'm still used to C++ with it's explicit return.

**Comment by NLincoln** [comment]
`test-tidy` issues have been fixed, `build` reports no errors, and I've cleaned up the comments a bit.

**Comment by tigercosmos** [comment]
I think there might be some result change when you run
```
./mach test-wpt tests/wpt/web-platform-tests/html/semantics/forms/the-input-element/
```
I expect at least `number.html` changed.

If you find some test case from expect Failed to Pass.
Delete the pair in `/Users/tigercosmos/servo/tests/wpt/metadata/html/semantics/forms/the-input-element/number.html.ini`

If all cases in `.ini` PASS, just delete the file. Other files are the same way.

File: components/script/dom/bindings/str.rs [hunk 3]
```diff
@@ -303,6 +303,35 @@ impl DOMString {
...
+                if val.is_nan() {
+                    return false;
+                }
+                if input.ends_with(".") {
+                    return false;
+                }
+                if input.starts_with("+") {
+                    return false;
+                }
+                true
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: These lines of `if` expressions can actually be all simplified to:
```rust
!(val.is_infinite() || val.is_nan() || input.ends_with('.') || input.starts_with('+'))
```

File: components/script/dom/bindings/str.rs [hunk 4]
```diff
@@ -303,6 +303,35 @@ impl DOMString {
         parse_week_string(&*self.0).is_ok()
     }
 
+    /// A valid number is the same as what rust considers to be valid,
+    /// except for +1., NaN, and Infinity.
+    /// https://html.spec.whatwg.org/multipage/#valid-floating-point-number
+    pub fn is_valid_number_string(&self) -> bool {
+        let input = &self.0;
+        match input.parse::<f64>() {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: This can actually benefit from using `map_or` as well:
```rust
input.parse::<f64>().map_or(false, |val| {
    // conditions...
})
```

**Review by KiChjang — CHANGES_REQUESTED** [review]
Great work! I have some small nits here, fix those, and you can r=me :)

**Comment by NLincoln** [comment]
I've made the changes :+1:

**Comment by KiChjang** [comment]
Alright! Next step is to squash your commits, and I'll happily land your PR :)

**Comment by NLincoln** [comment]
Done! I also rebased it onto master :smile:

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks for your contribution!

**Comment by KiChjang** [comment]
@bors-servo retry
* Gecko media
